common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala

```diff
@@ -132,13 +133,54 @@ object PipelineStorageStream extends Logging {
     msgSender: MessageSender[MsgDestination],
     indexer: Indexer[T])(implicit
                          ec: ExecutionContext,
-                         indexable: Indexable[T]) =
-    Flow[(Message, Option[T])]
-      .collect { case (msg, Some(document)) => Bundle(msg, document) }
+                         indexable: Indexable[T]) = {
+    val maxSubStreams = Integer.MAX_VALUE
+    Flow[(Message, List[T])]
+      .collect {
+        case (msg, items @ _ :: _) =>
+          items.map(item =>
+            Bundle[T](message = msg, item = item, numberOfItems = items.size))
+      }
+      .mapConcat[Bundle[T]](identity)
       .via(batchIndexFlow(config, indexer))
-      .mapAsyncUnordered(config.parallelism) {
-        case Bundle(msg, id) =>
-          Future.fromTry(msgSender.send(id).map(_ => msg))
+      .via(groupByMessage(maxSubStreams, 5 minutes).mergeSubstreams)
+      .mapConcat(identity)
+      .mapAsyncUnordered(config.parallelism) { bundle =>
+        for {
+          _ <- Future.fromTry(msgSender.send(indexable.id(bundle.item)))
+        } yield bundle
+      }
+      .via(groupByMessage[T](maxSubStreams, 5 minutes).collect {
+        case head :: _ => head.message
+      }.mergeSubstreams)
+  }
+
+  // Splits the flow into a substream for each messageId.
+  // Each substream emits one message with the complete list of bundles for the same messageId
+  // or no message if it didn't receive the correct number of bundles
+  def groupByMessage[T](maxSubStreams: Int, t: FiniteDuration) =
```

Comment:
I think it would be easier to read this help having the full function signature including output type here

Reply:
I can add it if you like, but the result signature is a bit weird cause it's a subflow. I think it's more confusing than anything but if you think it clarifies then sure

Reply:
It was mainly so I could see the output type, leave it out if you think really bad

Reply:
I've added it, let me know what you think. I don't find a way to simplify it

Reply:
Scratch everything I said...this can output a flow and the signature becomes much better